docs/proposals/secure-impersonation.md

```diff
@@ -0,0 +1,636 @@
+Alternatively, we could use [CRD defaulting](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting), but this puts other considerations on the design with detecting empty values.
+
+#### kubeConfig
+Kustomization and HelmRelease already have a `kubeConfig` option in addition to `serviceAccountName` and the proposed `user`.
+KubeConfig's are loaded from Secrets in the same Namespace in the style of Cluster API.
+They are safe to use literally as their own credential, since the FluxUser can be granted access to it as opposed to problematically assuming a controller SA.
+Not all kubeConfigs may be assumed to have impersonation rights, but there is no field that explicitly disables User impersonation.
+When `user`/`serviceAccountName` is unset and `kubeConfig` is non-empty, we could decide to ignore impersonation, and use the literal credential.
+Impersonation of Users and ServiceAccounts as well as ServiceAccount Token Impersonation can still function with an explicitly set value and does work across cluster boundaries.
+An impersonated User for the remote cluster will always match the Namespace used for the reconciler in the source cluster.
+
+KubeConfigs that mention the controller SA token in `/var/run/secrets/kubernetes.io/serviceaccount` must be rejected, because this is a privilege escalation route.
+These filepaths need to be resolved to absolute paths before they are checked to prevent subversive path traversal via symlinks.
+
+KubeConfigs that set `users[].user.exec.command` or `users[].user.auth-provider.config.cmd-path` should be rejected unless all `[command|cmd-path]` values return
+an [`os.exec.LookPath()`](https://golang.org/pkg/os/exec/#example_LookPath) inside the `/kubeconfig-bin/` of the controller Pod filesystem.
+While executing the kubeConfig, the `$PATH` likely needs to be preserved in case the auth-helper needs to execute other POSIX binary dependencies.
+Restricting `[command|cmd-path]` to `/kubeconfig-bin/` allows the administrator an opt-in directory where they can add tools they trust/accept like `gcloud` and `aws-iam-authenticator` 
+while disallowing access to more dangerous binaries like `kubectl`. This mechanism is somewhat inspired by `/sbin` and `setuid` binaries.
+
+Using exec helpers is not an approach used by Cluster API.
+In addition to being insecure, it's not possible to distrubite controllers containing all of the necessary binaries to support all kubernetes platforms.
+Cluster API providers instead schedule regular updates to kubeConfig Secret data which is more tenant friendly and dependency-free.
+
+Regardless, we must sanitize the kubeConfig, so it's easy to add a restricted, PATH-based allow-list that the flux-system owner can control if they choose to manage remote clusters in this way.
+
+
+#### sourceRefs
+Controllers should impersonate the reconciler User when fetching sourceRefs.
```

Comment:
Did you consider the effect of removing cross-namespace references? It would still be necessary to impersonate; but perhaps easier to reason about.